### PR TITLE
Return new node on modification

### PR DESCRIPTION
**Pull request by CoderZhi**
Return a new node on any modification. With this change, it is possible to reuse the nodes and minimize the spending in constructing a trie.

https://github.com/iotexproject/iotex-core/issues/2847

### CODE REVIEW

**Comment by Liuhaai** [comment]
Could an new pr be created to introduce `client` interface?

**Comment by CoderZhi** [comment]
> Could an new pr be created to introduce `client` interface?

is it hard to understand client interface?

File: db/trie/mptrie/hashnode.go [hunk 1]
```diff
@@ -8,53 +8,52 @@ package mptrie
 
 type hashNode struct {
 	node
-	mpt     *merklePatriciaTrie
 	hashVal []byte
 }
 
-func newHashNode(mpt *merklePatriciaTrie, ha []byte) *hashNode {
-	return &hashNode{mpt: mpt, hashVal: ha}
+func newHashNode(_ client, ha []byte) *hashNode {
```
Comment (dustinxie) [inline db/trie/mptrie/hashnode.go:14]: why leave `client` here if it's not used?

File: db/trie/mptrie/hashnode.go [hunk 2]
```diff
@@ -8,53 +8,52 @@ package mptrie
...
+	return &hashNode{hashVal: ha}
 }
 
-func (h *hashNode) Flush() error {
+func (h *hashNode) Flush(_ client) error {
```
Comment (dustinxie) [inline db/trie/mptrie/hashnode.go:18]: same

File: db/trie/mptrie/hashnode.go [hunk 3]
```diff
@@ -8,53 +8,52 @@ package mptrie
...
 }
 
-func (h *hashNode) loadNode() (node, error) {
-	return h.mpt.loadNode(h.hashVal)
+func (h *hashNode) loadNode(cli client) (node, error) {
+	return cli.loadNode(h.hashVal)
 }
 
-func (h *hashNode) Hash() ([]byte, error) {
+func (h *hashNode) Hash(_ client) ([]byte, error) {
```
Comment (dustinxie) [inline db/trie/mptrie/hashnode.go:57]: same here

File: db/trie/mptrie/sortedlist.go [hunk 4]
```diff
@@ -6,31 +6,40 @@ import (
 
 // SortedList is a data structure where elements are in ascending order
 type SortedList struct {
-	li       []uint8
-	isSorted bool // lazy-initilization
+	li     []uint8
+	sorted bool
```
Comment (dustinxie) [inline db/trie/mptrie/sortedlist.go:10]: does not hurt to leave the comment

File: db/trie/mptrie/node.go [hunk 5]
```diff
@@ -17,20 +17,28 @@ var ErrNoData = errors.New("no data in hash node")
 type (
 	keyType []byte
 
+	client interface {
```
Comment (dustinxie) [inline db/trie/mptrie/node.go:20]: seems to me `merklePatriciaTrie` implements this interface, is there more purpose/use of this interface?

if that's the case, name it like `trieOps`? honestly `client` is a bit confusing in this context

File: db/trie/mptrie/node.go [hunk 6]
```diff
@@ -17,20 +17,28 @@ var ErrNoData = errors.New("no data in hash node")
...
-		Search(keyType, uint8) (node, error)
-		Delete(keyType, uint8) (node, error)
-		Upsert(keyType, uint8, []byte) (node, error)
-		Hash() ([]byte, error)
-		Flush() error
+		Search(client, keyType, uint8) (node, error)
+		Delete(client, keyType, uint8) (node, error)
+		Upsert(client, keyType, uint8, []byte) (node, error)
+		Hash(client) ([]byte, error)
+		Flush(client) error
```
Comment (dustinxie) [inline db/trie/mptrie/node.go:33]: is it possible not add `client` to the func's input, if the node struct (leaf, extension, branch) contains the `client` interface?
feel it's a bit heavy/clumsy with this extra input

File: db/trie/mptrie/node.go [hunk 7]
```diff
@@ -17,20 +17,28 @@ var ErrNoData = errors.New("no data in hash node")
...
 	serializable interface {
 		node
-		hash(flush bool) ([]byte, error)
-		proto(flush bool) (proto.Message, error)
-		delete() error
-		store() (node, error)
+		hash(client, bool) ([]byte, error)
+		proto(client, bool) (proto.Message, error)
+		delete(client) error
+		store(client) error
```
Comment (dustinxie) [inline db/trie/mptrie/node.go:41]: same here

File: db/trie/mptrie/branchnode.go [hunk 8]
```diff
@@ -199,48 +212,78 @@ func (b *branchNode) child(key byte) (node, error) {
...
 }
 
-func (b *branchNode) updateChild(key byte, child node, hashnode bool) (node, error) {
-	if err := b.delete(); err != nil {
+func (b *branchNode) updateChild(cli client, key byte, child node, hashnode bool) (node, error) {
+	if err := b.delete(cli); err != nil {
 		return nil, err
 	}
+	var indices *SortedList
+	var children map[byte]node
```
Comment (Liuhaai) [inline db/trie/mptrie/branchnode.go:233]: the creation of `children`  could be moved before if condition
```
children := make(map[byte]node, len(b.children))
for k, v := range b.children {
   children[k] = v
}
if child == nil {
   delete(b.children, key)
   .......
} else {
   children[key] = child
}
```

File: db/trie/mptrie/sortedlist.go [hunk 9]
```diff
@@ -6,31 +6,40 @@ import (
...
 		li = append(li, k)
 	}
 	return &SortedList{
-		li:       li,
-		isSorted: false,
+		li: li,
 	}
 }
 
+func (sl *SortedList) sort() {
```
Comment (Liuhaai) [inline db/trie/mptrie/sortedlist.go:30]: let's keep private func below public func

**Comment by Liuhaai** [comment]
Other things seems good for me apart from two things: 1. how is the performance of new `func (b *branchNode) updateChild()` compared with old one in production; 2. I am not very confident about the change in sync mode. If a full sync could be done to verify the change, that'd be better.

File: db/trie/mptrie/branchnode.go [hunk 10]
```diff
@@ -199,48 +212,72 @@ func (b *branchNode) child(key byte) (node, error) {
...
 	}
-	_, err := b.store()
-	return err
+
+	return b.store(cli)
 }
 
-func (b *branchNode) updateChild(key byte, child node, hashnode bool) (node, error) {
-	if err := b.delete(); err != nil {
+func (b *branchNode) updateChild(cli client, key byte, child node, hashnode bool) (node, error) {
```
Comment (huof6829) [inline db/trie/mptrie/branchnode.go:228]: `hashnode` is not used

File: db/trie/mptrie/extensionnode.go [hunk 11]
```diff
@@ -159,52 +159,30 @@ func (e *extensionNode) commonPrefixLength(key []byte) uint8 {
...
-			return hn, nil
-		}
-	}
-	return e, nil
+	return newExtensionNode(cli, path, e.child)
 }
 
-func (e *extensionNode) updateChild(newChild node, hashnode bool) (node, error) {
-	err := e.delete()
+func (e *extensionNode) updateChild(cli client, newChild node, hashnode bool) (node, error) {
```
Comment (huof6829) [inline db/trie/mptrie/extensionnode.go:180]: `hashnode` is not used

**Comment by raullenchai** [comment]
PTAL @dustinxie @Liuhaai @huof6829

**Comment by Liuhaai** [comment]
Except from `hashnode bool` mentioned by @huof6829, other things look good for me.

**Review by huof6829 — APPROVED** [review]

**Review by dustinxie — APPROVED** [review]

**Pull request merged**